xxxx-generic-js-field.md

```diff
@@ -0,0 +1,170 @@
+
+```ocaml
+module Js = Js_of_ocaml.Js
+
+(* interfaces that snarky expects and returns *)
+module type Field_intf = sig
+  type t
+
+  val add : t -> t -> t
+
+  val sub : t -> t -> t
+
+  val mul : t -> t -> t
+
+  val zero : t
+
+  val one : t
+end
+
+module type Backend_intf = sig
+  module Field : Field_intf
+end
+
+module type Snarky_intf = sig
+  module Field : Field_intf
+
+  module Constraint : sig
+    type t
+
+    val r1cs : Field.t -> Field.t -> Field.t -> t
+
+    val equal : Field.t -> Field.t -> t
+  end
+
+  val assert_ : Constraint.t -> unit
+end
+
+module Snarky (Backend : Backend_intf) = struct
+  module Field = Backend.Field
+
+  module Constraint = struct
+    type t = unit
+
+    let r1cs _x _y _z : t = ()
+
+    let equal _x _y : t = ()
+  end
+
+  let assert_ (_c : Constraint.t) : unit = ()
+end
+
+(* generic native field module, reimplemented in JS *)
+type bigint = Js.Unsafe.any Js.t
+
+type field = bigint ref
+
+type js_field_module =
+  < add : (field -> field -> field) Js.readonly_prop
+  ; sub : (field -> field -> field) Js.readonly_prop
+  ; mul : (field -> field -> field) Js.readonly_prop
+  ; zero : field Js.readonly_prop
+  ; one : field Js.readonly_prop >
+  Js.t
+
+external create_js_field : field -> js_field_module = "create_field"
+
+(* make existing snarky-bindings a functor from snarky *)
+module Snarky_bindings (Impl : Snarky_intf) = struct
+  module Field' = struct
+    let assert_equal x y = Impl.assert_ (Impl.Constraint.equal x y)
+
+    let assert_mul x y z = Impl.assert_ (Impl.Constraint.r1cs x y z)
+  end
+end
+
+(* instantiating snarky and snarky-bindings with JS field module, and return something usable in JS *)
+let create_field (modulus : field) =
+  (* call generic JS field constructor *)
+  let js_field = create_js_field modulus in
+  (* instantiate backend *)
+  let module Backend = struct
+    module Field = struct
+      type t = field
+
+      let add = js_field##.add
+
+      let sub = js_field##.sub
+
+      let mul = js_field##.mul
+
+      let zero = js_field##.zero
+
+      let one = js_field##.one
+    end
+  end in
+  (* instantiate snarky *)
+  let module Snarky' = Snarky (Backend) in
+  (* instantiate snarky-bindings *)
+  let module Snarky_bindings' = Snarky_bindings (Snarky') in
+  (* return for JS use *)
+  object%js
+    method assertEqual = Snarky_bindings'.Field'.assert_equal
+
+    method assertMul = Snarky_bindings'.Field'.assert_mul
+  end
+```
+
+### Rust side: instantiating core Kimchi methods
+
+TODO - does Rust enable us to dynamically instantiate traits (or whatever) as function return values?
+If not, we might have to create Kimchi bindings for a _fixed number of supported provers_ (e.g. Pasta/IPA, BN128/KZG) and in a JS wrapper dynamically pick one of them that matches the field configuration given to `instantiateWithCustomField()`.
+
+## Test plan and functional requirements
+
+TODO
+
+## Drawbacks
+
+It's work.
+
+## Rationale and alternatives
```

Comment:
Other alternatives at the API level (though I like your proposal a lot!)

1. Add a type parameter to the `Field` class: ie. `Field<Bn128>` or `Field<Pasta>`
2. Dynamically change the behavior of Field at runtime: `Field.setBackendTo(Bn128())`
3. Statically instantiate the backends and tie that to the name of the class: `FieldBn128` or the import `import Field from 'lib/field_bn128'` <-- we can consider the latter as a shorthand built on top of the more general system you're proposing.

Reply:
Yeah I considered 2) but it's not very nice. 3) without a generic e2e solution is a reasonable fallback in case we dlcan't make fully generic work (parts of the logic would still be generic).
I don't think 1) fully specifies an API because types in TS don't affect the runtime, so there would still have to be some runtime step of picking the field